src/helper.jl

```diff
@@ -59,7 +59,7 @@ mutable struct XpressProblem <: CWrapper
         ptr = ref[]
         @assert ptr != C_NULL "Failed to create XpressProblem. Received null pointer from Xpress C interface."
         p = new(ptr, Any[], 0.0)
-        finalizer((p) -> destroyprob(p), p)
+        atexit(() -> destroyprob(p))
```

Comment:
if I understand correctly, `atexit` is only called at the `exit` command, while `finalizer` is called at garbage collection.
Correct me if I am wrong.
If I am right, we will acumulate problems during execution and only get rid of them when the program finishes, this would lead to huge memory issues.

Reply:
I meant to add a `TODO` comment here. 

Yes, you are correct that functions registered through `finalizer` are called when the objects are garbage collected. However, functions registered with [`atexit`](https://docs.julialang.org/en/v1/base/base/#Base.atexit) are run before object finalizers. This means that if we use `finalizer` in the struct constructor, we cannot register the [`free` function to run during `atexit`](https://github.com/JuliaOpt/Xpress.jl/blob/7bb86ec04036a07697dcd934ecb646685b0dec28/src/Xpress.jl#L90) when we call `Xpress.initialize()`. If we do use `finalizer` in the inner constructor to call `destroyprob` AND use `atexit` during `initialize` to call `free`, then:

```
julia --project
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.3.0-rc4.1 (2019-10-15)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> using Xpress
[ Info: Precompiling Xpress [9e70acf3-d6c9-5be6-b5bd-4e2c73e3e054]

julia> Xpress.initialize()
[ Info: Xpress: Found license file /Users/USER/local/xpressmp/bin/xpauth.xpr
[ Info: Xpress: Development license detected.

julia> xp = Xpress.XpressProblem()
Xpress Problem:
    type   : LP
    sense  : minimize
    number of variables                    = 0
    number of linear constraints           = 0
    number of quadratic constraints        = 0
    number of sos constraints              = 0
    number of non-zero coeffs              = 0
    number of non-zero qp objective terms  = 0
    number of non-zero qp constraint terms = 0
    number of integer entities             = 0


julia> exit()
ERROR - call to XPRS function when global environment not initialised - please call XPRSinit() before attempt to call Xpress-Optimizer functions.
error in running finalizer: Xpress.XpressError(errorcode=255, msg="Unable to call destroyprob")
```

we will get the above error during `exit()`.

I posted this question on slack and @simonbyrne suggested I look at [MathLink.jl](https://github.com/JuliaInterop/MathLink.jl/blob/7624db2d977990659f310103f2c7bf7d0266aa23/src/init.jl) and [MPI.jl](https://github.com/JuliaParallel/MPI.jl/blob/f471b62fea9ff46411f83556b95eac51023a9667/src/environment.jl) to see how they manage environments, and I think we can do something similar.

Reply:
I never had any problem by not freeing an xpress env, did you?

Reply:
I have not, but I figured it exists for a reason and thought I should do it. I can experiment with an implementation that has both `destroyprob` and `free` and if I'm not able to come up with something that works and is clean, I'll just use what is in the current `master`, i.e. not use `free` and register `destroyprob` using `finalizer`.

Reply:
nice, we leave the `free` thing as a TODO